core/src/env/api.rs

```diff
@@ -12,127 +12,439 @@
-use ink_prelude::vec::Vec;
+use ink_primitives::Key;
+
+/// Returns the address of the caller of the executed contract.
+///
+/// # Errors
+///
+/// If the returned caller cannot be properly decoded.
+pub fn caller<T>() -> Result<T::AccountId>
+where
+    T: EnvTypes,
+{
+    <EnvInstance as OnInstance>::on_instance(|instance| TypedEnv::caller::<T>(instance))
+}
 
-/// Stores the given value under the specified key in the contract storage.
+/// Returns the transferred balance for the contract execution.
 ///
-/// # Safety
+/// # Errors
 ///
-/// This operation is unsafe because it does not check for key integrity.
-/// Users can compare this operation with a raw pointer dereferencing in Rust.
-pub unsafe fn store(key: Key, value: &[u8]) {
-    ContractEnvStorage::store(key, value)
+/// If the returned value cannot be properly decoded.
+pub fn transferred_balance<T>() -> Result<T::Balance>
+where
+    T: EnvTypes,
+{
+    <EnvInstance as OnInstance>::on_instance(|instance| {
+        TypedEnv::transferred_balance::<T>(instance)
+    })
 }
 
-/// Clears the data stored at the given key from the contract storage.
+/// Returns the current price for gas.
 ///
-/// # Safety
+/// # Errors
 ///
-/// This operation is unsafe because it does not check for key integrity.
-/// Users can compare this operation with a raw pointer dereferencing in Rust.
-pub unsafe fn clear(key: Key) {
-    ContractEnvStorage::clear(key)
+/// If the returned value cannot be properly decoded.
+pub fn gas_price<T>() -> Result<T::Balance>
+where
+    T: EnvTypes,
+{
+    <EnvInstance as OnInstance>::on_instance(|instance| {
+        TypedEnv::gas_price::<T>(instance)
+    })
 }
 
-/// Loads the data stored at the given key from the contract storage.
+/// Returns the amount of gas left for the contract execution.
 ///
-/// # Safety
+/// # Errors
 ///
-/// This operation is unsafe because it does not check for key integrity.
-/// Users can compare this operation with a raw pointer dereferencing in Rust.
-pub unsafe fn load(key: Key) -> Option<Vec<u8>> {
-    ContractEnvStorage::load(key)
+/// If the returned value cannot be properly decoded.
+pub fn gas_left<T>() -> Result<T::Balance>
+where
+    T: EnvTypes,
+{
+    <EnvInstance as OnInstance>::on_instance(|instance| TypedEnv::gas_left::<T>(instance))
 }
 
-/// Returns the given data back to the caller.
+/// Returns the current block time in milliseconds.
+///
+/// # Errors
+///
+/// If the returned value cannot be properly decoded.
+pub fn block_timestamp<T>() -> Result<T::Timestamp>
+where
+    T: EnvTypes,
+{
+    <EnvInstance as OnInstance>::on_instance(|instance| {
+        TypedEnv::block_timestamp::<T>(instance)
+    })
+}
+
+/// Returns the account ID of the executed contract.
 ///
 /// # Note
 ///
-/// This operation must be the last operation performed by a called
-/// smart contract before it returns the execution back to its caller.
-pub fn return_data<T, E>(data: T)
+/// This method was formerly known as `address`.
```

Comment:
:prince: :guitar:

Reply:
? :sweat_smile:

Reply:
https://www.google.com/search?client=ubuntu&hs=VkJ&channel=fs&q=Why+is+Prince+called+the+Artist+Formerly+Known+as+Prince%3F&sa=X&ved=2ahUKEwj8vtP1tLXnAhWVonEKHQtBA_IQzmd6BAgREAw&biw=1720&bih=1273

Reply:
xD